### ISSUE TITLE
"Permission denied" as non-root user after creating/editing files as root

### ISSUE THREAD

**Issue by Monsieur-Moony** [termux/termux-app#335]
**Steps to reproduce**
1. Start up Termux and ensure you are in /data/data/com.termux/files/home/
2. Run `su`
3. Create a file with `touch test` or edit and save a file using vim
4. `exit` out of su
5. Run `rm test` or try to read file with vim
6. Change file permissions and ownership as root and try step 5 again

**Expected behaviour**
File can be deleted/read/edited as non-root user.

**Observed**
"Permission denied" occurs on any attempt to manipulate the file, including copying.

**Comments**
I tried `chmod 777` on first the file, and then on home/, and then tried `chown <user>` with no success. This bug doesn't happen in ~/storage/downloads.

I observed this error after editing my .bashrc in vim as a root user. After this, my non-root user does not even load up .bashrc when logging into bash.

**Comment by Monsieur-Moony** [comment]
After updating Termux today, I can't reproduce the issue. Closing now ...

**Comment by Monsieur-Moony** [comment]
Update: This bug has cropped up again.

**Comment by aleph4** [comment]
I encountered the same issue and did some investigation.

This is a SELinux security context issue:

```ShellSession
taimen:/data/data/com.termux/files/usr/etc # touch test
taimen:/data/data/com.termux/files/usr/etc # ls -lZ test
-rw-r--r-- 1 root    root    u:object_r:app_data_file:s0              0 2017-12-09 12:06 test
```

```ShellSession
taimen:/data/data/com.termux/files/usr/etc # chown u0_a158:u0_a158 test
taimen:/data/data/com.termux/files/usr/etc # ls -lZ test
-rw-r--r-- 1 u0_a158 u0_a158 u:object_r:app_data_file:s0              0 2017-12-09 12:06 test
```

and can be resolved by running restorecon on the affected file:

```ShellSession
taimen:/data/data/com.termux/files/usr/etc # restorecon -v test
SELinux: Loaded file_contexts
SELinux:  Relabeling /data/data/com.termux/files/usr/etc/test from u:object_r:app_data_file:s0 to u:object_r:app_data_file:s0:c512,c768.
taimen:/data/data/com.termux/files/usr/etc # ls -lZ test
-rw-r--r-- 1 u0_a158 u0_a158 u:object_r:app_data_file:s0:c512,c768    0 2017-12-09 12:06 test
```

**Comment by Monsieur-Moony** [comment]
Thanks, that fixed it! 👍
I wonder if it would be helpful to have a FAQ for issues, because I imagine many people will encounter this issue.

**Comment by Grimler91** [comment]
@Monsieur-Moony We have a FAQ at [the termux wiki](https://wiki.termux.com/wiki/FAQ), but this issue isn't covered (feel free to contribute).

**Comment by Monsieur-Moony** [comment]
@Grimler91 Done! Feel free to review [my edit](https://wiki.termux.com/wiki/FAQ#I_updated_termux_packages_in_root_and_now_I.27m_getting_.27permission_denied.27_as_a_non-root_user) and let me know if I can improve it.

**Comment by Muhaiminkp** [comment]
@aleph4 
I am this issue but the command restorecon tells 

bash-4.4# restorecon -Rv /data/data/com.termux/files/usr/etc/                             selinux_android_file_context_handle: Error getting file context handle (No such file or directory)

Help any one

**Comment by its-pointless** [comment]
you have to move files to /sdcard at which point the files aren't marked by selinux. Then mv them back to where you need them.

**Comment by jottr** [comment]
Where can I find this `restorecon`-package?   
`pkg search restorecon` doesn't return anything, nor does `$ restorecon`.

**Comment by Monsieur-Moony** [comment]
Strange. I have 2 Android devices (one rooted and one not), and they both have (toybox)[http://landley.net/toybox/about.html] installed on them, which contains packages like restorecon, getenforce etc for SELinux support.

I can only guess that either your device doesn't have `/system/bin/toybox` installed, or it does but toybox was compiled without SELinux support.

**Comment by jottr** [comment]
Misunderstanding on my part: I overread that it's not in termux's `PATH` but can be invoked by giving the full path `/system/bin/restorecon`, which works on my tmux installation as well.

**Comment by Monsieur-Moony** [comment]
Interesting, thanks for sharing the solution!

**Comment by xeffyr** [comment]
> Could this way work

$cd /data/data/com.termux
$restorecon -v -R
$cd /data/app/com.termux-1/
$restorecon -v -R

to do with all files of termux??

Running restorecon on Termux directory **/data/data/com.termux** shouldn't harm anything. It just restores normal SELinux labels for files at this path.

**Comment by Justsoos** [comment]
@xeffyr  I have run 
```
# restorecon -v -R /data/data/com.termux                                                                                                       
SELinux: Loaded file_contexts contexts from /file_contexts.
SELinux:  Relabeling /data/data/com.termux/lib from u:object_r:app_data_file:s0 to u:object_r:app_data_file:s0:c512,c768.
SELinux:  Relabeling /data/data/com.termux/files/home/.python_history from u:object_r:app_data_file:s0 to u:object_r:app_data_file:s0:c512,c768.
SELinux:  Relabeling /data/data/com.termux/files/home/.bash_history from u:object_r:app_data_file:s0 to u:object_r:app_data_file:s0:c512,c768.
SELinux:  Relabeling /data/data/com.termux/files/home/.viminfo from u:object_r:app_data_file:s0 to u:object_r:app_data_file:s0:c512,c768.
SELinux:  Relabeling /data/data/com.termux/files/usr/lib/python3.6/__pycache__/rlcompleter.cpython-36.pyc from u:object_r:app_data_file:s0 to u:object_r:app_data_file:s0:c512,c768.
SELinux:  Relabeling /data/data/com.termux/files/usr/var/lib/dpkg/info/zsh.list from u:object_r:app_data_file:s0 to u:object_r:app_data_file:s0:c512,c768.
SELinux:  Relabeling /data/data/com.termux/files/usr/var/lib/dpkg/status-old from u:object_r:app_data_file:s0 to u:object_r:app_data_file:s0:c512,c768.
```
it corrected some files I had not found. 
now `pkg install XXX` works well, but, after such tossing, I can not run `reboot` command in termux with command not found prompt, as I remember, it can be run as non-root before.

**Comment by JonasT** [comment]
Since the wiki is quite lacking on this and suggests to just run `restore`con which doesn't work from a normal termux shell, here is what to do on a rooted device to fix the SELinux context of a file (or do anything else that needs the proper SELinux tools around & working):

1. Open termux
2. Get root with `tsu`
3. Use this command to get a proper rooted system shell (the native shell with the SELinux tools, not termux):
   ```
   env -i USER=root "$(PATH=/system/xbin:/system/bin:/su/bin:/sbin:/magisk/.core/bin which su)" root --context u:r:shell:s0 --shell /system/bin/sh --command sh
   ```

   (if your device wasn't rooted with magisk, some paths may need to be adjusted)

4. Inside the system shell, `cd` to the directory with your file with wrong permissions
5. Now you can use `restorecon -v <affectedfile>`. Don't forget to check the result with `ls -lZa` and also fix the owner & permissions as necessary with `chown` and `chmod`! (e.g. if this is a file that should be used by another app located in that app's folder of which you messed up the permissions, then it needs to be owned by that app and read/executable or whatever the app requires on access)

**Comment by xoihiox** [comment]
Ran into this same issue.  As stated, this has *nothing* to do with file ownership.  I guess I accidentally installed one or more packages with sudo.  Here is how I solved the issue (fingers crossed):
1. `sudo dpkg -P` all packages that were installed with sudo.
2. Activate a root shell with `su -l root'
3. run restorecon on all offending files, namely `/data/data/com.termux/files/usr/var/lib/dpkg/status`
4. I had to manually track down and delete a number of installed components as apt update spit out similar errors.  
5. I ran `restorecon -v -R` from  `/data/data/com.termux` multiple times and nothing broke, so it may be possible just to go with that command, though I still had to track down and delete stuff.

Similar to above, but it may be necessary to delete and purge all components of packages installed with sudo.  Several of these operations will require sudo, which meant I had to execute number `3` above multiple times to get `pkg` to `update` successfully.
